### PR TITLE
DeploymentInitializeOptions spec update

### PR DESCRIPTION
**Pull request by DrusTheAxe**
Update to the DeploymentManager spec with new initialization options to support the DeploymentManager auto-initializer. 

DeploymentManager's auto-initializer is modeled after the Bootstrapper's auto-initializer as they pose similar behavioral requirements.

For more information see

* [DeploymentManager auto-initializer #2835](https://github.com/microsoft/WindowsAppSDK/pull/2835) and for more details.
* [Updated Bootstrapper spec for MddBootstrapIntialize2(...options) #2335](https://github.com/microsoft/WindowsAppSDK/pull/2335)
* [Rich information on Bootstrap initalization failure #2316](https://github.com/microsoft/WindowsAppSDK/pull/2316)

https://task.ms/41000657

### CODE REVIEW

**Review by sachintaMSFT — APPROVED** [review]

File: specs/Deployment/DeploymentAPI.md [hunk 1]
```diff
@@ -221,6 +266,26 @@ namespace Microsoft.Windows.ApplicationModel.WindowsAppRuntime
...
+
+        /// If not successful call DebugBreak() if a debugger is attached to the process
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_DebugBreak_IfDebuggerAttached;
+
+        /// If not successful perform a fail-fast
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_FailFast;
+
+        /// If not successful show UI
```
Comment (FrankC-msft) [inline specs/Deployment/DeploymentAPI.md:282]: The order of the options in DeploymentInitializeOptions definition doesn't match the order they're applied in the examples section.  I assume the order of application is deterministic.  This should be clear in the API details or comments in this section.

File: specs/Deployment/DeploymentAPI.md [hunk 2]
```diff
@@ -221,6 +266,26 @@ namespace Microsoft.Windows.ApplicationModel.WindowsAppRuntime
...
+
+        /// If not successful call DebugBreak()
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_DebugBreak;
+
+        /// If not successful call DebugBreak() if a debugger is attached to the process
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_DebugBreak_IfDebuggerAttached;
+
+        /// If not successful perform a fail-fast
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:278]: They're not mutually exclusive and, per API guidelines, "bit-fields for multiple properties" went out of vogue some years back in favor of "multiple boolean properties" (better developer experience and evolutionary growth).

File: specs/Deployment/DeploymentAPI.md [hunk 3]
```diff
@@ -221,6 +266,26 @@ namespace Microsoft.Windows.ApplicationModel.WindowsAppRuntime
...
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_FailFast;
+
+        /// If not successful show UI
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_ShowUI;
+
+        /// Do nothing (do not error) if the process lacks package identity
+        [contract(DeploymentContract, 3)]
+        Boolean OnNoPackageIdentity_NOOP;
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:288]: 'Empty' doesn't quite convey it. Lingo lately is "packaged app" = has package identity and "unpackaged app" = lacks package identity.

Expressing it code-wise `if (not process.HasPackageIdentity) { ignore/skip/return }`. Coming up with a good phrase...

* OnNoPackageIdentity
* OnLacksPackageIdentity
* OnUnpackagedProcess

Nothing really struck me as all that wonderful. Any hot flashes of Eureka! by you?

File: specs/Deployment/DeploymentAPI.md [hunk 4]
```diff
@@ -179,12 +179,57 @@ they shut down, to refer to the updated framework package.
...
+        hr = _Initialize(...)
+    }
+    else
+    {
+        // We're not a packaged process. Is calling Initialize an error or ignore it?
+        if options.OnNoPackageIdentity_NOOP
+            return OK
+        else
+            hr = HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED)
+    }
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:205]: Should have `{}`. Initially wrote pseudo-code and got more involved so I turned it into real code. Forgot to fully do that.

RECOMMEND: Add `{}` for if+else bodies, per C/C++ Coding Standards

File: specs/Deployment/DeploymentAPI.md (hunk 1 above)
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:282]: The behavior will be doc'd (needs to be). Reordering physical order of the properties isn't a substitute for that, is it?

File: specs/Deployment/DeploymentAPI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrankC-msft) [reply specs/Deployment/DeploymentAPI.md:278]: I thought we were back on the "flags" side of the pendulum again, thanks for the context :)

Not mixing independent with related options is a good callout, that should usually be avoided in either flags or bool format.  Maybe the two "DebugBreak" should be combined into one non-flag { Never, IfDebuggerAttached, Always } enum as a peer of the FailFast and ShowUI booleans.

File: specs/Deployment/DeploymentAPI.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (FrankC-msft) [reply specs/Deployment/DeploymentAPI.md:288]: AllowUnpackagedDeployment ?   Getting the right lingo for Packaged/Identity is important but that isn't what feels unnatural to me.  This is an "InitializeOption", so I think we can express it in terms of what the developer is signing up for, instead of an implicit event in the initialize code.  ("OnNoPackageIdentity", On* sounds like an 'event' occurring but it's actually an initial condition of the deployment, right?)

File: specs/Deployment/DeploymentAPI.md [hunk 5]
```diff
@@ -179,12 +179,57 @@ they shut down, to refer to the updated framework package.
...
+Additional behavior is available if an error occurs in `DeploymentManager.Initialize()`,
+per the following logic:
+
+```c++
+DeploymentManager::Initialize(...)
+{
+    HRESULT hr = S_OK
+
+    // Only supported in packaged processes
+    if IsPackagedProcess()
```
Comment (jonwis) [inline specs/Deployment/DeploymentAPI.md:193]: Parens after "if"

File: specs/Deployment/DeploymentAPI.md [hunk 6]
```diff
@@ -179,12 +179,57 @@ they shut down, to refer to the updated framework package.
...
+            return OK
+        else
+            hr = HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED)
+    }
+    if FAILED(hr)
+    {
+        LogToEventLog(hr, ...)
+
+        // Should we pop the debugger?
+        if options.OnError_DebugBreak or (options.OnError_DebugBreak_IfDebuggerAttached and IsDebuggerPresent())
```
Comment (jonwis) [inline specs/Deployment/DeploymentAPI.md:211]: This looks like C++ but isn't. Should this instead point to the implementation? This code-oriented description of the behavior is hard to follow.

Feels like there should be a sample of showing the app setting these flags.

File: specs/Deployment/DeploymentAPI.md [hunk 7]
```diff
@@ -221,6 +266,26 @@ namespace Microsoft.Windows.ApplicationModel.WindowsAppRuntime
         /// WindowsAppSDK main and singleton packages will be shut down forcibly if they are
         /// currently in use, when registering the WinAppSDK packages.
         Boolean ForceDeployment;
+
+        /// If not successful call DebugBreak()
+        [contract(DeploymentContract, 3)]
+        Boolean OnError_DebugBreak;
```
Comment (jonwis) [inline specs/Deployment/DeploymentAPI.md:272]: We don't use underscores in names.

File: specs/Deployment/DeploymentAPI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonwis) [reply specs/Deployment/DeploymentAPI.md:278]: This is asking for an "ErrorActions" enum:

```c#
enum DeploymentErrorActions {
    ReportError = 0, // default, normal 'return oops'
    Break,
    BreakInDebugger,
    FailFast,
    DisplayErrorUI,
}
```

Feels like you should only be able to specify one action, not some combination.  The most common case is likely "just report the error" and the others are all some form of "how do I want this to act during debugging?"

File: specs/Deployment/DeploymentAPI.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonwis) [reply specs/Deployment/DeploymentAPI.md:288]: Why is this even an option here? The app developer shouldn't have to set this _and_ build an MSI, they just ... build the MSI.  What's this bit even trying to convey or avoid?

File: specs/Deployment/DeploymentAPI.md (hunk 5 above)
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:193]: RECOMMEND: Real code or English logic, to avoid confusion

File: specs/Deployment/DeploymentAPI.md (hunk 6 above)
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:211]: RECOMMNED: Revise per comment above

File: specs/Deployment/DeploymentAPI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:278]: RECOMMEND: Remove BreakInDebugger + FailFast

[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:278]: RECOMMEND: BOOL ShowUI + (always) DebugBreak (no option/API)

File: specs/Deployment/DeploymentAPI.md (hunk 1 above)
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:282]: DISCUSSION: Only 1 option now so this comment falls out / NA

File: specs/Deployment/DeploymentAPI.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:288]: RECOMMEND: Drop NoOp option and make behavior "DoNothingIfUnpackaged" i.e. NoOp option is always TRUE

File: specs/Deployment/DeploymentAPI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:278]: There's a problem with the recommendation regarding what to do when errors occur

TL;DR proposed change is to add
`enum OnErrorBreak { None, DebugBreak }`
at a minimum or preferably
`enum OnErrorBreak { None, DebugBreak, DebugBreakIfDebuggerAttached, FailFast }`


The proposed logic was
```
class AutoInitializer
{
public:
    AutoInitializer()
    {
        DeploymentInitializeOptions options{ GetOptions() }
        HRESULT hr{ DeploymentManager.Initialize(options) }
        IF FAILED(hr)
        {
            IF options.OnErrorShowUI
            {
                MessageBox(...)
            }
            IF options.OnErrorDebugBreak OR (IsDebuggerAttached AND options.OnErrorDebugBreakIfDebuggerAttached)
            {
                DebugBreak()
            }
            IF options.OnErrorFailFast
            {
                FAIL_FAST_HR(hr)
            }
            ExitProcess(hr)
        }
    }
};
AutoInitializer g_autoInitializer;

```

API Review feedback changed that `IF FAILED` block to

```
IF FAILED(hr)

{
    IF options.OnErrorShowUI
    {
        MessageBox(...)
    }
    DebugBreak()
    ExitProcess(hr)
}

```

The summary of changes

1. Drop support for FailFast

2. Drop support for DebugBreakIfDebuggerAttached

3. Make DebugBreak unconditional

#3 is a problem as it griefs end user scenarios


Some of the scenarios driving the initial proposal

* User runs installed app

* Developer in VS hits F5

* Developer testing app in headless environment

These call for different behaviors

* ShowUI. No crash/break

* ShowUI + crash/break

* No UI. No crash if that causes user interaction to be needed (e.g. DebugBreak but AeDebug is set). FailFast provides a 'crash' w/o possibility of hang.

**PROBLEM#1:** Per API Review the debugbreak is unconditional. Spikes the end user experience.
 **This is a problem**

**SOLUTION#1:** Make debugbreak conditional



**PROBLEM#2?** Per API Review `DebugBreak` is the only 'crash' option. Is this a problem for automated environments e.g. does DebugBreak (in, say, a headless session-0 process) lead to process block/hang?

I still think the original options are worthy given our experiences with the Bootstrapper. Yes, being the one on the sharp-end-of-the-support-stick here for the past year I might have strong opinions on this subject... It's also consistent with the Bootstrapper shipped in WinAppSDK 1.1.  Do folks still feel "don't be so complicated" is the right answer?

**SOLUTION#2:** Support FailFast + DebugBreakIfDebuggerAttached


**In terms of what these mean to the API**


Per API Review we have 1 boolean

`bool OnErrorShowUI`


To solve #1 we could add a second boolean but better to make it an enum

```
bool OnErrorShowUI

enum OnErrorBreak { None, DebugBreak }

```


To solve #2 we extend the enum

```
bool OnErrorShowUI

enum OnErrorBreak { None, DebugBreak, DebugBreakIfDebuggerAttached, FailFast }
```

[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply specs/Deployment/DeploymentAPI.md:278]: RECOMMEND: Keep API per API Review BUT change DebugBreak to be conditional IFF a debugger is attached:

```
IF FAILED(hr)
{
    IF options.OnErrorShowUI
    {
        MessageBox(...)
    }
    IF IsDebuggerAttached()
    {
        DebugBreak()
    }
    ExitProcess(hr)
}
```

Developers can use [IFEO](https://docs.microsoft.com/previous-versions/windows/desktop/xperf/image-file-execution-options) if they desire other behavior e.g. to run an executable under a debugger whenever it's launched (w/o having to explicitly run the debugger and run the executable from it).
